server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java

```diff
@@ -170,48 +170,47 @@
      */
     volatile ReplicationGroup replicationGroup;
 
-    private final Map<String, RetentionLease> retentionLeases = new HashMap<>();
-
-    private Collection<RetentionLease> copyRetentionLeases() {
-        assert Thread.holdsLock(this);
-        return Collections.unmodifiableCollection(new ArrayList<>(retentionLeases.values()));
-    }
+    private RetentionLeases retentionLeases = RetentionLeases.EMPTY;
 
     /**
      * Get all non-expired retention leases tracked on this shard. An unmodifiable copy of the retention leases is returned. Note that only
      * the primary shard calculates which leases are expired, and if any have expired, syncs the retention leases to any replicas.
      *
      * @return the retention leases
      */
-    public Collection<RetentionLease> getRetentionLeases() {
+    public RetentionLeases getRetentionLeases() {
         final boolean wasPrimaryMode;
-        final Collection<RetentionLease> nonExpiredRetentionLeases;
+        final RetentionLeases nonExpiredRetentionLeases;
         synchronized (this) {
             if (primaryMode) {
                 // the primary calculates the non-expired retention leases and syncs them to replicas
                 final long currentTimeMillis = currentTimeMillisSupplier.getAsLong();
                 final long retentionLeaseMillis = indexSettings.getRetentionLeaseMillis();
-                final Collection<RetentionLease> expiredRetentionLeases = retentionLeases
+
+                final Map<String, RetentionLease> leases = RetentionLeases.toMap(retentionLeases);
+
+                final Collection<RetentionLease> expiredRetentionLeases = leases
```

Comment:
Since we have made the map non-modifiable and pushed it to RetentionLeases, this flow might be simpler if we collect non-expired leases instead?

Reply:
I think that we will want to collect both so that we can easily check if any leases are expired. I'll push a change showing what I mean.

Reply:
@dnhatn I pushed 9d0998962f3f0b47879ce212ab65e7b191320381

Reply:
ah, is there any issue if we compare the count of non-expired leases to the count of the current leases? I might miss something.

Reply:
That would be fine too but I prefer the explicitness of this. Let me know what you prefer.

Reply:
I am fine with your choice :).